src/Chainweb/Pact/PactService.hs

```diff
@@ -573,90 +598,143 @@ validateHashes bHeader pData miner transactions =
+                    <> "; current hash: " <> blockHashToText (_blockHash cur)
+                    <> "; current height: " <>  sshow (_blockHeight cur)
+                    <> "; checkpointer hash: " <> blockHashToText ph
+                    <> "; checkpointer height: " <>  sshow h
+
+            parent <- ParentHeader
+                <$!> lookupBlockHeader ph (T.pack caller <> ".syncParentHeader")
+            setParentHeader (caller <> ".syncParentHeader") parent
+            return parent
+
+-- | INTERNAL FUNCTION. ONLY USE WHEN YOU KNOW WHAT YOU DO!
+--
+-- Same as 'withCheckpointer' but doesn't rewinds the checkpointer state to
+-- the provided target. Note that it still restores the state to the target.
+--
+-- In other words, this method can move the checkpointer back in time to a state
+-- in the current history. It doesn't resolve forks or fast forwards the
 -- checkpointer.
 --
--- Usually, one needs to rewind the checkpointer first to the target. In those
--- cases the function 'withCheckpointerRewind' should be preferred.
+-- /NOTE:/
 --
--- The result of the inner action indicates whether the resulting checkpointer
--- state should be discarded or saved.
+-- In most use cases one needs to rewind the checkpointer first to the target
+-- and function 'withCheckpointer' should be preferred.
 --
--- If the inner action throws an exception the checkpointer state is discarded.
+-- Only use this function when 1. you need the extra performance from skipping
+-- the call to 'rewindTo' and 2. you know exactly what you do.
 --
-withCheckpointer
+withCheckpointerWithoutRewind
     :: PayloadCasLookup cas
     => Maybe (BlockHeight, BlockHash)
-        -- The current block height and the parent hash
+        -- ^ The block height @height@ to which to restore and the parent header
+        -- @parent@.
+        --
+        -- It holds that @(_blockHeight parent == pred height)@
+
     -> String
+        -- ^ Putative caller
     -> (PactDbEnv' -> PactServiceM cas (WithCheckpointerResult a))
     -> PactServiceM cas a
-withCheckpointer target caller act = mask $ \restore -> do
-    cenv <- restore $ restoreCheckpointer target caller
-    try (restore (act cenv)) >>= \case
-        Left e -> discardTx >> throwM @_ @SomeException e
-        Right (Discard !result) -> discardTx >> return result
-        Right (Save header !result) -> saveTx header >> return result
+withCheckpointerWithoutRewind target caller act = do
+    checkPointer <- getCheckpointer
+    logInfo $ "restoring (with caller " <> caller <> ") " <> sshow target
+
+    -- check requirement that this must be called within a batch
+    unlessM (asks _psIsBatch) $
+        error $ "Code invariant violation: withCheckpointer called by " <> caller <> " outside of batch. Please report this as a bug."
+    -- we allow exactly one nested call of 'withCheckpointer', which is used
+    -- during fastforward in 'rewindTo'.
+    unlessM ((<= 1) <$> asks _psCheckpointerDepth) $ do
```

Comment:
I'm not satisfied with the implementation of the two new invariants. We can consider merging this PR in interests of expedience as I am satisfied it fixes the issues, but would be happier if we resolve the following issues:

1. `psIsBatch` is only enforced here, which means that we only care about it when we're doing checkpointer restore/finalize. We need to consider whether batching is even meaningful in PactService as a separate operation -- the question being, *do we always pair batch calls with a `withCheckpointerXXX` call*? If so we should just do the batching here and get rid of the flag.

2. If `psIsBatch` is truly necessary as a separate invariant, we should consider enforcing against nested batching in `withBatchXXX` as I'm pretty sure we don't want that. 

2. `psCheckpointerDepth` semantically indicates that we never want to nest `withCheckpointerXXX` calls. Thus it should be a boolean, and cannot be `True` coming into this function.

Reply:
These invariants can and should both be enforced on the type level. We never expect them to trigger in production. It's just for the case that maintainers of the code forget to nest transactions properly --  which obvious is easy to miss :-)

Implementing it this way, was the least intrusive (smallest) change I could think of. But I fully agree that this should be replaced by an implementation that supports proper scoping more explicitly.

1. Agreed, but I considered it out of scope for this PR to change that. So, given that this invariant exists and that it's unnoticed violation caused bugs, I added this check as a temporary solution to enforce this invariant.

2. Yes, we can add that as a third invariant, too. Also, I don't think anything would break if someone nests it. But not 100%.

3. I can't be Boolean, since it is actually called nested. But only 2 levels deep. This check ensures that we don't accidentally introduce additional nesting. It's a bit like a golden test. If a programer would change the nesting structure, they would be notified here. So it prevents accidentally messing things up.

Reply:
Actually, in the first iteration I used `Bool` and only then noticed that there are nested calls :-)

Reply:
Oh I see, `0` and `1` are valid states. Yuck ... when is it nested?

Reply:
In `rewindTo` blocks are replayed to fast forward the checkpointer. Each block runs in it's own `BLOCK` SAVEPOINT environment. That the only place where it's nested. I put this check to avoid that we accidentally add another nesting. I would be particularly concerned about recursive nesting.